lib/spack/spack/solver/asp.py

```diff
@@ -0,0 +1,1838 @@
+        self.backend.add_rule(
+            [],
+            [atoms[s] for s in symbols] +
+            [-atoms[s] for s in negated_symbols]
+            + rule_atoms
+        )
+
+    def iff(self, expr1, expr2):
+        self.rule(head=expr1, body=expr2)
+        self.rule(head=expr2, body=expr1)
+
+    def one_of_iff(self, head, versions):
+        # if there are no versions, skip this one_of_iff
+        if not versions:
+            return
+
+        self.out.write("%s :- %s.\n" % (head, AspOneOf(*versions)))
+        self.out.write("%s :- %s.\n" % (AspOneOf(*versions), head))
+
+        at_least_1_sym = fn.at_least_1(*head.args).symbol()
+        at_least_1 = self.backend.add_atom(at_least_1_sym)
+
+        more_than_1_sym = fn.more_than_1(*head.args).symbol()
+        more_than_1 = self.backend.add_atom(more_than_1_sym)
+
+        version_atoms = [self.backend.add_atom(f.symbol()) for f in versions]
+        self.backend.add_weight_rule(
+            [at_least_1], 1, [(v, 1) for v in version_atoms])
+        self.backend.add_weight_rule(
+            [more_than_1], 2, [(v, 1) for v in version_atoms])
+
+        head_atom = self.backend.add_atom(head.symbol())
+        self.backend.add_rule([head_atom], [at_least_1, -more_than_1])
+
+        self.backend.add_rule([], [head_atom, more_than_1])
+        self.backend.add_rule([], [head_atom, -at_least_1])
+
+    def solve(
+            self, solver_setup, specs, dump=None, nmodels=0,
+            timers=False, stats=False, tests=False
+    ):
+        timer = Timer()
+
+        # Initialize the control object for the solver
+        self.control = clingo.Control()
+        self.control.configuration.solve.models = nmodels
+        self.control.configuration.asp.trans_ext = 'all'
+        self.control.configuration.asp.eq = '5'
+        self.control.configuration.configuration = 'tweety'
+        self.control.configuration.solve.parallel_mode = '2'
+        self.control.configuration.solver.opt_strategy = "usc,one"
+
+        # set up the problem -- this generates facts and rules
+        self.assumptions = []
+        with self.control.backend() as backend:
+            self.backend = backend
+            solver_setup.setup(self, specs, tests=tests)
+        timer.phase("setup")
+
+        # read in the main ASP program and display logic -- these are
+        # handwritten, not generated, so we load them as resources
+        parent_dir = os.path.dirname(__file__)
+        self.control.load(os.path.join(parent_dir, 'concretize.lp'))
+        self.control.load(os.path.join(parent_dir, "display.lp"))
+        timer.phase("load")
+
+        # Grounding is the first step in the solve -- it turns our facts
+        # and first-order logic rules into propositional logic.
+        self.control.ground([("base", [])])
+        timer.phase("ground")
+
+        # With a grounded program, we can run the solve.
+        result = Result()
+        models = []  # stable moodels if things go well
```

Comment:
```suggestion
        models = []  # stable models if things go well
```
unless I'm missing a joke?

Reply:
Thanks!